xds/googledirectpath/googlec2p.go

```diff
@@ -0,0 +1,219 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/google"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/googlecloud"
+	internalgrpclog "google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/resolver"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+	"google.golang.org/grpc/xds/internal/client/bootstrap"
+	"google.golang.org/grpc/xds/internal/env"
+	"google.golang.org/grpc/xds/internal/version"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	_ "google.golang.org/grpc/xds" // To register xds resolvers and balancers.
+)
+
+const (
+	c2pScheme = "google-c2p"
+
+	tdURL          = "directpath-trafficdirector.googleapis.com"
+	httpReqTimeout = 10 * time.Second
+	zoneURL        = "http://metadata.google.internal/computeMetadata/v1/instance/zone"
+	ipv6URL        = "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s"
+
+	gRPCUserAgentName               = "gRPC Go"
+	clientFeatureNoOverprovisioning = "envoy.lb.does_not_support_overprovisioning"
+	ipv6CapableMetadataName         = "TRAFFICDIRECTOR_DIRECTPATH_C2P_IPV6_CAPABLE"
+
+	logPrefix = "[google-c2p-resolver]"
+)
+
+// For overriding in unittests.
+var (
+	onGCE = func() bool {
+		return googlecloud.OnGCE()
+	}
+	newClientWithConfig = func(config *bootstrap.Config) error {
+		_, err := xdsclient.NewWithConfig(config)
+		return err
+	}
+
+	childBuilders = map[string]resolver.Builder{
+		"dns": resolver.Get("dns"),
+		"xds": resolver.Get("xds"),
+	}
+
+	logger = internalgrpclog.NewPrefixLogger(grpclog.Component("directpath"), logPrefix)
+
+	defaultNode = &v3corepb.Node{
+		Id:                   "C2P",
+		Metadata:             nil, // To be set if ipv6 is enabled.
+		Locality:             nil, // To be set to the value from metadata.
+		UserAgentName:        gRPCUserAgentName,
+		UserAgentVersionType: &v3corepb.Node_UserAgentVersion{UserAgentVersion: grpc.Version},
+		ClientFeatures:       []string{clientFeatureNoOverprovisioning},
+	}
+
+	ipv6EnabledMetadata = &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			ipv6CapableMetadataName: {
+				Kind: &structpb.Value_BoolValue{BoolValue: true},
+			},
+		},
+	}
+)
+
+func init() {
+	resolver.Register(&c2pResolverBuilder{})
+}
+
+type c2pResolverBuilder struct{}
+
+// Build helps implement the resolver.Builder interface.
+func (b *c2pResolverBuilder) Build(t resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	ret := &c2pResolver{
+		target: t,
+		cc:     cc,
+		opts:   opts,
+	}
+	// start does I/O. Run it in a goroutine.
+	go ret.start()
+	return ret, nil
+}
+
+// Name helps implement the resolver.Builder interface.
+func (*c2pResolverBuilder) Scheme() string {
+	return c2pScheme
+}
+
+type c2pResolver struct {
+	target resolver.Target
+	cc     resolver.ClientConn
+	opts   resolver.BuildOptions
```

Comment:
Maybe just pass these around as parameters instead, since they aren't needed after construction?

Reply:
Constructor doesn't build the child. The build options are needed later when building the child.